{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0afee92f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***********\n",
      "K00548\n",
      "rhizo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K00651\n",
      "rhizo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K00799\n",
      "rhizo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K01505\n",
      "rhizo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K07160\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K15554\n",
      "rhizo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K22955\n",
      "rhizo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K00003\n",
      "rhizo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K00032\n",
      "rhizo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K00036\n",
      "rhizo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K00383\n",
      "rhizo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K02439\n",
      "phyllo & soil\n",
      "Pseudomonadales\n",
      "***********\n",
      "***********\n",
      "K00387\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K00432\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K00548\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K00640\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K05301\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K06048\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K15554\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n",
      "***********\n",
      "K17228\n",
      "phyllo & soil\n",
      "Rhizobiales\n",
      "***********\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Load data\n",
    "aggregated_tests = pd.read_excel('/Users/hanli/Desktop/FYP/PAML/aggregated_all_tests_combined.xlsx', sheet_name='remove_0_rows')\n",
    "final_matrix = pd.read_excel('/Users/hanli/Desktop/FYP/PAML/final_matrix_with_metadata.xlsx', sheet_name='metadata')\n",
    "genes_dir = '/Users/hanli/Desktop/FYP/PAML/genomes/kofamscan/csv'\n",
    "gff_dir = '/Users/hanli/Desktop/FYP/PAML/genomes/gff'\n",
    "fna_dir = '/Users/hanli/Desktop/FYP/PAML/genomes/fna'\n",
    "faa_dir = '/Users/hanli/Desktop/FYP/PAML/genomes/faa'\n",
    "ffn_dir = '/Users/hanli/Desktop/FYP/PAML/genomes/ffn'\n",
    "\n",
    "\n",
    "def extract_sequences(genome, gene, target_id):\n",
    "    ffn_file = os.path.join(ffn_dir, f'{genome}.ffn')\n",
    "    try:\n",
    "        with open(ffn_file) as ffn:\n",
    "            for record in SeqIO.parse(ffn_file, 'fasta'):\n",
    "                if target_id in record.id:\n",
    "                    return record\n",
    "            return None\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {ffn_file}\")\n",
    "        return\n",
    "\n",
    "\n",
    "def extract_aa_sequences(genome, gene, target_id):\n",
    "    faa_file = os.path.join(faa_dir, f'{genome}.faa')\n",
    "    try:\n",
    "        with open(faa_file) as faa:\n",
    "            for record in SeqIO.parse(faa_file, 'fasta'):\n",
    "                if target_id in record.id:\n",
    "                    return record\n",
    "            return None\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found: {faa_file}\")\n",
    "        return\n",
    "\n",
    "for _, row in aggregated_tests.iterrows():\n",
    "    if row['method'] == 'KOfamScan' and row['Order'] in ['Rhizobiales', 'Pseudomonadales'] and row['selected'] == 1:\n",
    "        gene = row['gene']\n",
    "        niche = row['niche']\n",
    "        order = row['Order']\n",
    "        niche_new = niche.replace(' & ', '_')\n",
    "        print(\"***********\")\n",
    "        print(gene)\n",
    "        print(niche)\n",
    "        print(order)\n",
    "        print(\"***********\")\n",
    "        # Filter genomes\n",
    "        if niche == \"rhizo & soil\":\n",
    "            spheres = [\"Rhizosphere\", \"Soil\"]\n",
    "        elif niche == \"phyllo & soil\":\n",
    "            spheres = [\"Phyllosphere\", \"Soil\"]\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "        selected_genomes = final_matrix[(final_matrix['sphere'].isin(spheres)) & (final_matrix['Order'] == order)]\n",
    "        \n",
    "        # Extract sequences\n",
    "        sequences = []\n",
    "        aa_sequences = []\n",
    "        for genome in selected_genomes['Gene']:\n",
    "            csv_file = os.path.join(genes_dir, f'{genome}.csv')\n",
    "            with open(csv_file) as csvfile:\n",
    "                for line in csvfile:\n",
    "                    if gene in line:\n",
    "                        identifier, _ = line.strip().split(',', 1)\n",
    "                        scaffold_id, cds_number = identifier.rsplit('_', 1)\n",
    "                        scaffold_id = scaffold_id.replace('\"', '')\n",
    "                        cds_number = cds_number.replace('\"', '')\n",
    "\n",
    "                        # Extract the corresponding target ID from the GFF file\n",
    "                        gff_file = os.path.join(gff_dir, f'{genome}.gff')\n",
    "                        \n",
    "                        # Define a custom exception for breaking out of nested loops\n",
    "                        class TargetIdFound(Exception):\n",
    "                            pass\n",
    "                        \n",
    "                        target_id = None  # Initialize target_id with a default value\n",
    "\n",
    "                        try:\n",
    "                            with open(gff_file) as gff:\n",
    "                                cds_list = []\n",
    "                                for line in gff:\n",
    "                                    if line.startswith(scaffold_id) and \"CDS\" in line:\n",
    "                                        cds_list.append(line)\n",
    "                                cds_number_int = int(cds_number)\n",
    "                                cds_list = [cds for cds in cds_list if \"CDS\" in cds]\n",
    "                                if cds_number_int > len(cds_list):\n",
    "                                    raise TargetIdFound()\n",
    "                                target_cds = cds_list[cds_number_int - 1]\n",
    "                                target_id = target_cds.split(\"ID=\")[1].split(\";\")[0]\n",
    "                                raise TargetIdFound()\n",
    "                        except TargetIdFound:\n",
    "                            pass\n",
    "\n",
    "                        # Proceed only if target_id is defined\n",
    "                        if target_id is not None:\n",
    "                            seq = extract_sequences(genome, gene, target_id)\n",
    "                            aa_seq = extract_aa_sequences(genome, gene, target_id)\n",
    "                            if seq is not None:\n",
    "                                sequences.append((genome, seq))\n",
    "                            if aa_seq is not None:\n",
    "                                aa_sequences.append((genome, aa_seq))\n",
    "                                # ... The previous code\n",
    "\n",
    "                                # Save nucleotide sequences to a FASTA file\n",
    "                                niche_order = f\"{niche.replace(' & ', '_')}_{order}\"\n",
    "                                nucleotide_output_file_path = f'/Users/hanli/Desktop/FYP/PAML/new_fna_faa/{gene}_{niche_order}_ffn_output.fasta'\n",
    "                                with open(nucleotide_output_file_path, 'w') as fasta_file:\n",
    "                                    for genome, record in sequences:\n",
    "                                        header = f\"{genome}_{record.id} {record.description.split(' ', 1)[1]}\"\n",
    "                                        fasta_file.write(f'>{header}\\n{record.seq}\\n')\n",
    "\n",
    "                                # Save amino acid sequences to a FASTA file\n",
    "                                aa_output_file_path = f'/Users/hanli/Desktop/FYP/PAML/new_fna_faa/{gene}_{niche_order}_faa_output.fasta'\n",
    "                                with open(aa_output_file_path, 'w') as fasta_file:\n",
    "                                    for genome, record in aa_sequences:\n",
    "                                        header = f\"{genome}_{record.id} {record.description.split(' ', 1)[1]}\"\n",
    "                                        fasta_file.write(f'>{header}\\n{record.seq}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6463a5ad",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
